src/source/CppGenerator.scala

```diff
@@ -91,11 +91,34 @@ class CppGenerator(spec: Spec) extends Generator(spec) {
   }
 
   def generateHppConstants(w: IndentWriter, consts: Seq[Const]) = {
+    var needsCppGen = false
+
     for (c <- consts) {
+      var isConstExpr = true
+      val constValue = c.value match {
```

Comment:
It might be clearer to match on c.ty instead.  I'm not set on that, since this does make the code simpler, but it took me a while to figure out that this distinction is the reason that optionals are being treated as constexpr-eligible.  It's not an explicit case statement anywhere, but simply implicit in the fact that the constant value for an optional<T> is of type T.